### ISSUE TITLE
How to just update status in homekit

### ISSUE THREAD

**Issue by cis2131** [mschm/node-red-contrib-homekit#76]
I am integrating with a home controll system, that sends status changes when a light turns on (from the physical switch) to the homekit node, but when the node recieves the change, it sends the value to the output. This results in a "loop" effect on the controller.
 
How can I just update homekit status, without sending the value out again.?

**Comment by julynessi** [comment]
a simple solution ..... just put "rbe node" before an mqtt output to prevent send duplicate msg (yes it's break the loop)

**Comment by maragelis** [comment]
Use this in a function node after the homekit node

if (msg.hap.context !== undefined )
{
return msg;
}

**Comment by cis2131** [comment]
Thank you, that works fine..

I have another issue.. 
If i send e.g. {"Brightness": 40 } to a LightBulb, then the node, registers it, but the button in homekit dont show the value. 
What am i doing wrong?.

**Comment by jonathanschneider** [comment]
An rbe node will only block a duplicate message from the HomeKit node. When the status update comes from outside the HomeKit node, the rbe node will still let the message from the HomeKit node (that's already a duplicate of the message from the source) pass. In that case, the source of the update will receive its update to the HomeKit node again from the HomeKit node. That will cause the HomeKit node to send a duplicate message which then will be blocked by the rbe node.
That solution only works if the status in the source doesn't change in between sending its update and receiving the same update again from the HomeKit node. If the status changes in between, it will create a loop nonetheless.
I, for example, have a Wi-Fi dimmer that I can control with buttons. If I dim using the buttons the Wi-Fi dimmer sends out several updates before it receives the first update from HomeKit and thus jumps back and forth between the current value from the button and an old value from HomeKit. In those cases evaluating the "msg.hap.context" object is better.

**Comment by gonzunigad** [comment]
It would be great if homekit accesory node has two inputs. One for stat update, and one for applying and update. Or maybe just passing through the same message as the input.

**Comment by gonzunigad** [comment]
For a Sonoff S20 (The switch) flashed with Tasmota, I did: 

![image](https://user-images.githubusercontent.com/1103494/41206809-6ef40d88-6cd8-11e8-83b4-a8bceed0f516.png)

### First
```javascript
msg.payload = {On: msg.payload == 'ON'};
return msg;
```

### Second function

```javascript
var lastState = context.get('lastState') || false;

if (typeof(msg.hap) == 'undefined') { //This comes from the MQTT state. In that case we just update our internal status flag. We do not send anything back to mqtt to prevent a loop

    context.set('lastState', msg.payload.On);
    return null;
}

if (msg.payload.On == context.get('lastState')) {
    return null;
}

context.set('lastState', msg.payload.On);
msg.payload = msg.payload.On ? 'ON' : 'OFF';   
return msg;
```

**Comment by frooonk** [comment]
Hi gonzunigad, I try to apply your solution with msg.payload.Brightness (numbers between 0...100) instead of msg.payload.On (boolean true/false). 

Can you give me a hint please?

Thank you very much!
Frank

![homekit_lightbulb_edit](https://user-images.githubusercontent.com/11030445/50112126-c05c1980-023e-11e9-9e33-1dd1d6811254.PNG)

**Comment by futuretec** [comment]
Hi @frooonk,

as stated in the documentation of this HomeKit node you can inject any invalid payload (e.g. `{"foo":"bar"}`) into your HomeKit device to get a debug log message with all allowed Characteristics.

e.g. for a Lightbulb the output should look like this:
>Characteristic foo cannot be written.
>Try one of these: Name, On, Brightness, Hue, Saturation, Name, ColorTemperature

I tried and succeeded by directly injection the following JSON payloads to the nodes:
* `{ "On" : true }`
* `{ "On" : false }`
* `{ "Brightness" : 50.0 }`
* `{ "Brightness" : 75 }`
(sending in the Brightness as float or int seems both to be working)

Changing the Brightness will emit an Change Event from the HomeKit node - but if you keep checking for an existing `hap.context` in the payload this should be fine:
```javascript
{
  "payload": {
    "Brightness": 50
  },
  "hap": {
    "oldValue": 0,
    "newValue": 50,
    "characteristic": {
      "displayName": "Brightness",
      "UUID": "00000000-0000-1000-0000-000000000001",
      "iid": null,
      "value": 50,
      "status": null,
      "eventOnlyCharacteristic": false,
      "props": {
        "format": "int",
        "unit": "percentage",
        "minValue": 0,
        "maxValue": 100,
        "minStep": 1,
        "perms": [
          "pr",
          "pw",
          "ev"
        ]
      },
      "subscriptions": 0,
      "_events": {},
      "_eventsCount": 1
    }
  },
  "name": "Lightbulb",
  "_msgid": "abcdefab.cdefab"
}
```

Greetings
Dominik

**Comment by frooonk** [comment]
Hi Dominik,

thanks for your help :-) I tried again and now I can display the Brightness status (from KNX) in HomeKit correctly !

But I still have one problem: everytime I change the Brightness setpoint in HomeKit (i.e. from 90% to 67%) the KNX-status jumps to 0%. See Debugging console in the new screenshot.

Here's my modified code of the node "HomeKit to KNX (Brightness)" inspired by gonzunigad. I hope I did it right.

`var lastState = context.get('lastState') || 0;if (typeof(msg.hap) == 'undefined') {`


![homekit loop_brightness_2](https://user-images.githubusercontent.com/11030445/50181448-b7d21480-030c-11e9-8112-3d95d658eed9.PNG)

What I am doing wrong ???

Thanks in a

**Comment by frooonk** [comment]
Hello, the problem is solved :-)

Problem:
Any change of the brightness setpoint in HomeKit will cause the HomeKit-node to emit a Change Event (Thanks to Dominik!). But with every (!) Change Event, the status "On" (true/false) of the HomeKit-node also changes. I guess this causes KNX to send the value "0" back. 

Solution:
Two rbe blocks are required so that the brightness is not set to value "0" with each change of the setpoint in Homekit (and thus the light goes out). See flow in the screenshot. 

Thank you again! 

Best regards, Frank

![flow1](https://user-images.githubusercontent.com/11030445/50560639-20ed5b00-0d03-11e9-934f-c365e6764ee6.PNG)
